core/network/impl/stream_engine.hpp

```diff
@@ -44,11 +47,40 @@ namespace kagome::network {
    private:
     struct ProtocolDescr {
       std::shared_ptr<ProtocolBase> protocol;
-      std::shared_ptr<Stream> incoming;
-      std::shared_ptr<Stream> outgoing;
-      std::queue<std::function<void(std::shared_ptr<Stream>)>>
+
+      struct {
+        std::shared_ptr<Stream> stream;
+      } incoming;
+
+      struct {
+        std::shared_ptr<Stream> stream;
+        bool reserved = false;
+      } outgoing;
+
+      std::deque<std::function<void(std::shared_ptr<Stream>)>>
           deffered_messages;
+
+      bool hasActiveOutgoing() const {
+        return outgoing.stream && !outgoing.stream->isClosed();
+      }
+
+      bool reserve() {
+        if (outgoing.reserved || hasActiveOutgoing()) return false;
+
+        outgoing.reserved = true;
+        return true;
+      }
+
+      void dropReserved() {
+        BOOST_ASSERT(outgoing.reserved);
```

Comment:
Should we call `outgoing.stream->reset()` here?

Reply:
No, we need only to drop the flag.